### PR TITLE
Better errors, verbosity for cargo commands, docs

### LINKED ISSUE #22: Better error handling and logging
Each error should have its own error code.
Logging should be configurable with --verbose and --silent

### LINKED ISSUE #27: Prepare docs for every public element

### PR DESCRIPTION
**Pull request by kubaplas**
Closes #22 
Closes #27

### CODE REVIEW

File: src/backend.rs [hunk 1]
```diff
@@ -1,36 +1,71 @@
...
 use crate::odra_dependency::odra_dependency;
 use crate::odra_toml::OdraConf;
-use crate::{command, consts, AddBackendCommand, RemoveBackendCommand};
+use crate::{command, consts, info, AddBackendCommand, RemoveBackendCommand};
 
+/// Type of dependency in Cargo.toml
 pub enum DependencyType {
+    /// `odra = { path = "../odra/core"}`
     Local,
+    /// `odra = { git = "https://github.com/odradev/odra" branch = "develop" }`
```
Comment (kpob) [inline src/backend.rs]:
```suggestion
    /// `odra = { git = "https://github.com/odradev/odra", branch = "develop" }`
```

File: src/backend.rs [hunk 2]
```diff
@@ -1,36 +1,71 @@
...
+impl Backend {
+    /// Returns the name of the backend. It is a name used internally by a project. Most of the time
+    /// it will be the same as package. It is useful, when you want to add more than one of the same
+    /// backends, for example in different versions, from github or one stored locally (e.g.
+    /// `casper`, `casper-local`, `casper-develop`).
+    pub fn name(&self) -> &String {
+        &self.name
+    }
+
+    /// Returns name of the dependency. For example `casper` will be later used to generate
```
Comment (kpob) [inline src/backend.rs]:
```suggestion
    /// Returns the name of the dependency. For example `casper` will be later used to generate
```

File: src/errors.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            Error::InvalidInternalCommand(_) => 2,
+            Error::FailedToReadCargo(_) => 3,
+            Error::NoBackendConfigured => 4,
+            Error::NoSuchBackend => 5,
+            Error::WasmTargetNotInstalled => 6,
+            Error::NotAnOdraProject => 7,
+        }
+    }
+
+    pub fn print_and_die(&self) -> ! {
```
Comment (kpob) [inline src/errors.rs:59]: `print_and_die()` :heart:

File: src/log.rs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    prettycli::warn(message);
+}
+
+/// Error message
+pub fn error(message: &str) {
+    prettycli::error(message);
+}
+
+/// Critical message
+pub fn _crit(message: &str) {
```
Comment (kpob) [inline src/log.rs:19]: Why does it start with an underscore?

File: src/backend.rs [hunk 5]
```diff
@@ -464,28 +226,215 @@ fn main() {}
...
+            conf.backends = Some(backends);
+            conf.save();
+            true
+        } else {
+            false
+        }
+    }
+
+    fn from_add_command(add: &AddBackendCommand) -> Backend {
+        let dependency;
```
Comment (kpob) [inline src/backend.rs]: You don't need to initialize `dependency` in every branch.
```suggestion
        let dependency = if ...
```

File: src/backend.rs [hunk 6]
```diff
@@ -464,28 +226,215 @@ fn main() {}
...
     }
 
-    /// Returns a path where builder lives
-    pub fn builder_path(&self) -> String {
-        format!(".builder_{}/", self.name())
+    /// Returns Odra dependency tailored for use by backend (optional set to true and wasm feature
+    /// enabled)
+    fn odra_dependency() -> Dependency {
+        let dependency = odra_dependency();
+        match dependency {
```
Comment (kpob) [inline src/backend.rs]:
```suggestion
        match odra_dependency() {
```

**Review by kpob — APPROVED** [review]

File: src/log.rs (hunk 4 above)
Comment (kubaplas) [reply src/log.rs:19]: It is not used yet, so clippy had some issues with it.

File: src/errors.rs (hunk 3 above)
Comment (kubaplas) [reply src/errors.rs:59]: <3

**Pull request merged**